pkg/sql/sqlliveness/slstorage/slstorage.go

```diff
@@ -498,10 +498,27 @@ func (s *Storage) Insert(
 	if err := s.txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
 		batch := txn.NewBatch()
 
-		k, _, err := s.keyCodec.encode(sid)
+		k, region, err := s.keyCodec.encode(sid)
+		if err != nil {
+			return err
+		}
+		unavailableRegions, err := s.livenessProber.QueryUnavailablePhysicalRegions(ctx, txn, true /*filterAvailable*/)
 		if err != nil {
 			return err
 		}
+		if unavailableRegions.ContainsPhysicalRepresentation(string(region)) {
+			// Delete all rows for the region in system.sqlliveness, system.sql_instances,
+			// system.leases.
+			if err := regionliveness.CleanupSystemTableForRegion(ctx,
+				s.codec, string(region), txn); err != nil {
+				return err
+			}
+			// Make the region as available again.
+			if err := s.livenessProber.MarkPhysicalRegionAsAvailable(ctx, txn, string(region), unavailableRegions[string(region)]); err != nil {
```

Comment:
The implementation of the recovery protocol looks correct to me. But there is one potential availability issue in the implementation I noticed while reviewing the RFC. 

The RFC contains this line:
```
If unvailable_at is not null, but set to a timestamp after the transaction timestamp, a SQL server can attempt to set it to null in a transaction modifying only the region_liveness table.
```

I think the reason I wrote that is I was assuming sql liveness would attempt to clear an `unavailable_at` set for its region before the region is marked as down. Currently, if any server sets `unavailable_at` the sql servers will all restart even if their region becomes available before the sessions would expire.

I believe it is safe to clear an `unavailable_at` without deleting any of the survive zone rows as long as the transaction deadline comes before the `unavailable_at` timestamp. 

This logic probably belongs in `Storage.Update` and should run whenever an unavailable at timestamp is observed and the sql server was able to read from the sqlliveness table.